parla/task_runtime.py

```diff
@@ -1225,23 +1254,54 @@ def run_mapper(self):
                 # move to the mapped task scheduling.
                 break
 
-    def run_deviceq_mapper(self):
+    def _schedule_tasks(self):
+        """ Currently this doesn't do any intelligent scheduling (ordering).
+            Dequeue all ready tasks and send them to device queues in order.
+        """
+        print("[Scheduler] Schedule Phase")
         while True:
             task: Optional[TaskBase] = self._dequeue_task()
             if not task or not task.assigned:
                 logger.debug("Task %r: Failed to assign", task)
                 break
             for d in task.req.devices:
-                print("[Scheduler] Task, ", task.name, ", is enqueued onto a device Q.", sep='')
-                worker = self._worker_threads[task.req.env_no]
-                worker._enqueue_task_local(task)
+                print(f"[Scheduler] Enqueuing {task.name} to device {d}")
+                self._device_queues[d].append(task)
+
+    def _launch_tasks(self):
+        """ Iterate through free devices and launch tasks on them
+        """
+        print("[Scheduler] Launch Phase")
+        with self._monitor:
+            for dev, queue in self._device_queues.items():
+                # Make sure there's an available WorkerThread
+                if len(self._free_worker_threads) == 0:
+                    break
+
+                # Hack to make launching work for now
+                # Only works for single-device tasks, and only one task at a time on each device
+                # TODO: Make this not terrible
+                if self._available_resources._occupancy_dict[dev] == False: # if not occupied
+                    if len(queue) > 0: # if there are tasks on the queue
+                        task = queue.pop() # grab a task
+                        worker = self._free_worker_threads.pop() # grab a worker
+                        print(f"[Scheduler] Launching {task.name} on WorkerThread {worker.index}")
+                        self._available_resources._occupancy_dict[dev] = True # mark the device as occupied
+                        worker.assign_task(task) # assign the task to the worker (this notifies the worker's monitor)
+                        print(f"[Scheduler] Launched {worker.task.name}")
 
     def run(self) -> None:
         # noinspection PyBroadException
         try: # Catch all exception to report them usefully
+            i = 0
             while self._should_run:
-                self.run_mapper()
-                self.run_deviceq_mapper()
+                self._map_tasks()
+                self._schedule_tasks()
+                self._launch_tasks()
```

Comment:
One thing I'm concerned about reading this is overhead and/or stalling. 

iiuc (if I understand correctly) the execution is: 
- A group of tasks are mapped to devices
- A group of tasks are scheduled (assigned priority in device queues corresponding to their mapping) and are ready to run
- All devices are iterated over, ready tasks are grabbed, and attempted to be run on worker threads. A worker thread can only have 1 task assigned to it (the currently running task). 

Assuming the above understanding, and that I'm not missing any work stealing, if any workers finish their work they need to wait until the next time L1300 is called to have any work to do. This can be after potentially expensive and time consuming "Map" and "Schedule" phases (depending on policy).

Reply:
On the other end, if this loop is very running quickly, then the "mapping" and "scheduling" are likely never populated and batches aren't considered at a time. (although this can be fixed by having a threshold maybe? Not sure what the best pattern is here)

Reply:
You understood correctly, IIUC and this worker thread pool causes GIL overheads compared to before. But the main point of the thread pool is to support multi-device task. If one thread per one device is, it is unclear how to run multi-devices task. If the GIL is the real problem, we need to revisit our previous dicussion about C++ back-end. 

Mapping and scheduling would not be populated. Threshold of length looks good to me. 

Your points make sense to me. I will list them on the sheet.

@sestephens73  Do you agree? or any suggestion?

Reply:
@dialecticDolt You're correct. This is why we've split the scheduler into three distinct phases: mapping, scheduling, and launching. In the (near) future when we're discussing policies, we can easily run the phases in different periods or loops or orders, etc. Launch can be set to run with a fast period while schedule and map have slower ones, for example. I think this would address your concerns?

@nicelhc13 
> this worker thread pool causes GIL overheads compared to before

I don't think so...how come?

Reply:
@sestephens73  After I think about this more, I am now thinking that it would not be an overhead compared to before.

In the past (when we have one thread per device Q), scheduler distributes tasks onto device Qs, each device Q has a dedicated thread, and they pop one head task.

Now we have thread pool and threads on the thread pool will try to get task.

I thought that in thread pool case, threads will compete to get task, it would be one more data race or phase, and therefore, it could cause more GIL invocations.

But this task is distributed through one scheduler thread by thread indexing, which is similar with a dedicated thread.

Sorry!